KAFKA-13024

Gerrrr opened the pull request:
**Disclaimer**: This patch is WIP. Its purpose is to demonstrate a chosen approach and get feedback from committers.a 

### Summary 

This patch solves [KAFKA-13024](https://issues.apache.org/jira/browse/KAFKA-13024) and[ KAFKA-13183](https://issues.apache.org/jira/browse/KAFKA-13183) by introducing a concept of `decorator nodes` in the topology. The idea behind decorator nodes is similar to a decorator pattern - to extract common added behaviour from different entities. 

Specifically, this patch introduces 2 decorator nodes - `DropNullKeyNode` and `DropNullKeyValueNode`. These nodes have a single method `decorate` that accepts a `ProcessorSupplier` and returns another `ProcessorSupplier` that filters `null` keys/key-values respectively. Besides fixing 2 issues mentioned above (dropping `null` keys on repartition, missing metrics), there are 2 advantages to this approach:
* It reduces code duplication around `if (record.key() == null...`. 
* As @mjsax suggested in [this comment](https://issues.apache.org/jira/browse/KAFKA-13024?focusedCommentId=17427868&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17427868), It optimizes away duplicated null-key/value processing from the topology.

### Implementation details

The decorators are set by nodes according to a new parameter in processor parameters:

https://github.com/apache/kafka/blob/b4c3adc3ce3f40ad96400b94379fa206d782922f/streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorParameters.java#L45-L52

Currently ± working example is `StatefulProcessorNode` where the original processor supplier is decorated: 

https://github.com/apache/kafka/blob/b4c3adc3ce3f40ad96400b94379fa206d782922f/streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StatefulProcessorNode.java#L85-L92

Decorators are added to the topology together with the decorating node:

https://github.com/apache/kafka/blob/b4c3adc3ce3f40ad96400b94379fa206d782922f/streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java#L229-L240

Then, `InternalStreamsBuilder#buildAndOptimizeTopology` removes duplicating downstream decorators and attaches the remaining ones to the decorated nodes. The decorator nodes are not present in the final topology.

Added tests in `InternalStreamBuilderTest` and `KStreamRepartitionIntegrationTest.java` describe how the decorator operations are done. 

---

The following description reflects the first take at this issue and is left here only for completeness.

~Right now, the `repartition` operator always filters out `null` keys. This behavior is not correct and is a regression compared to the deprecated `through` operator.~

~This patch fixes the issue by filtering `null` keys only for optimizable repartition nodes. First, it removes unnecessary processor node from the topology for `UnoptimizableRepartitionNode`. Second, it only adds the filtering processor only if the node is optimizable. I introduced an `isOptimizable` boolean method for repartition nodes to make this property more obvious in the code.~

~To verify that change, I added a new test to the repartition integration suite that produces pairs with `null` keys and expects them to be present after the repartition.~

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Gerrrr committed:
KAFKA-13024: Use not-null filter only in optimizable repartitions

Gerrrr commented:
CI test failure is caused by a a known flaky test - ([KAFKA-8541](https://issues.apache.org/jira/browse/KAFKA-8541]), so probably unrelated.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/GroupedTableOperationRepartitionNode.java:
```diff
@@ -119,5 +119,10 @@ public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
+
+        @Override
+        public boolean isOptimizable() {
+            return false;
```
Why is a table repartitioning not optimizable?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/UnoptimizableRepartitionNode.java:
```diff
@@ -53,19 +53,13 @@ private UnoptimizableRepartitionNode(final String nodeName,
     public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
         topologyBuilder.addInternalTopic(repartitionTopic, internalTopicProperties);
 
-        topologyBuilder.addProcessor(
```
Not sure if I understand this change? Can you elaborate?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KStreamRepartitionIntegrationTest.java:
```diff
@@ -672,6 +672,42 @@ public void shouldGoThroughRebalancingCorrectly() throws Exception {
+        builder.stream(inputTopic, Consumed.with(Serdes.Integer(), Serdes.String()))
+            .selectKey((key, value) -> value == null ? null : key.toString())
+            .repartition(repartitioned)
+            .mapValues(value -> value != null ? "mapped-" + value  : "default-value")
```
Why do we need the `mapValues()` step in this test?

mjsax reviewed (commented):
Not sure if I fully understand the change yet (would need to double check the code, but might be easier to just ask):
- is this PR removing the optimization for all cases when `repartition()` is uses (I think the answer is yes)?
- if yes, I would propose to keep the optimization if possible, e.g., `stream.repartition().groupByKey().aggregate()`: because the downstream operator would drop records with `null`-key on read (in fact also records with `null`-value on read for this example), we could push this filter upstream
- the tricky part is to determine when it is save to push the filter upstream:
```
KStream s1 = stream.repartition();
s1.map().to();
s1.groupByKey().aggregate();
```
For this example, it's not ok to add the optimization because not all downstream operators do the drop `null` key/value thing.

Not sure if it would be out of scope to follow this proposal, but the idea is to add the upstream filter if all downstream operators allow for it (ie, downstream is an aggregation or join).

Not sure if we also want to add the drop `null`-value optimization (maybe better in a follow up PR if we do), because it's more complex when we want to drop records based on `null`-values compared to `null`-keys.

Gerrrr committed:
Revert "KAFKA-13024: Use not-null filter only in optimizable repartitions"

This reverts commit e875e95aae4827f9fe8a68a624959f10fdcaaa19.

Gerrrr committed:
Repartition drops null keys only if all downstream operations drop them

Gerrrr commented:
Thanks for the review! I agree that solving this issue by removing the not-null optimization completely is very much suboptimal. 8f176cb reverts the first commit. 7c6e97f introduces another approach where `UnoptimizableRepartitionNode` adds the not-null filter iff all downstream operations drop the null keys.

Gerrrr committed:
WIP: Extract dropping null keys and values into a decorator node

The idea of this commit is to extract dropping records with null keys
and/or values into a separate upstream filter node. The main source of
all introduced complexity is to avoid altering the result topology.

This commit introduces DecoratorGraphNodes that change behavior of
the graph node they are prepended to, but do not appear on the end
topology. There are kinds of decorator nodes right now - DropNullKeyNode
and DropNullKeyValueNode. These nodes allow InternalStreamsBuilder to
move these filtering operations as upstream as possible.

This is a WIP commit. TODO:
* Finish work on joins
* Javadocs

Gerrrr commented:
@suhas-satish Good idea! I'll take some time next week to rebase and provide proper description.

guozhangwang commented:
Hello @Gerrrr @mjsax thanks for kicking starting on this issue! I made a first pass on it and have a few quick thoughts.

1) First I want to make sure I understand the scope of the fix. My take is that we are only fixing `KStream`, since for `KTable` it's always safe to filter null keys; and again it's trying to maintain the original optimization but only when all downstream operators after the repartition are going to filter null keys, i.e. as long as one downstream operator is not a per-key stateful operator that filters null keys then we should not apply this optimization. Is my understanding right?

2) If 1) is right, then I'm wondering if we can just use the `KStreamFilter` still, but only at the end of the topology building as we've constructed and learned about all the nodes in the topology, hence in the `optimize` function then. My main rationale is that, we would not be able to decide whether the optimization on the upstream can be used until the topology has been fully constructed hence only then we would learn about all the downstream nodes.

Gerrrr commented:
Thanks for taking a look at this PR @guozhangwang !

>  we are only fixing KStream, since for KTable it's always safe to filter null keys

Yes

>  it's trying to maintain the original optimization but only when all downstream operators after the repartition are going to filter null keys, i.e. as long as one downstream operator is not a per-key stateful operator that filters null keys then we should not apply this optimization.

This patch drops `null` keys on repartition only if the downstream operation drops them (see `KStreamRepartitionIntegrationTest#shouldDropNullKeysOnRepartitionWithDownstreamAggregate`).

First, we mark certain operations that filter `null` keys or keys and values as such via corresponding `ProcessorParameters` (see https://github.com/apache/kafka/pull/11447/files#diff-b80820b46c327e9a393d3b2435f3c141745d7c99b02196744dbe2bbdb2fd95acR65-R67). Based on these parameters, we prepend decorator nodes in the logical topology before their graph nodes (see `InternalStreamsBuilder#addGraphNode`). Then, `InternalStreamsBuilder#applyDecorators` traverses the topology and removes redundant decorators (e.g. multiple operations in a row filter our `null` keys). Finally, `InternalStreamsBuilder` removes decorators from the logical topology and attaches them to their successor nodes.

When `ProcessorGraphNodes` call `writeToTopology`, they check that a decorator is attached. If the decorator is attached, it wraps the processing code (e.g. https://github.com/apache/kafka/pull/11447/files#diff-61f97d0e2f4db6ed1fd994da154c3955fe943c76559217e27b0b1e10daba4623R85-R92)

See `InternalStreamsBuilder#applyDecorators` for more details.

> Can we just use the KStreamFilter?

@mjsax was concerned that adding `KStreamFilter` will change the physical topology and might break user workflows if they rely on node names. This patch does not introduce new nodes to the physical topology.

> we would not be able to decide whether the optimization on the upstream

I might be missing the full picture here as I am a bit new to the project, but, as I see it, the algorithm I described above does not suffer from this issue. There, we essentially extract `if (record.key() == null || record.value() == null) {` code from the operations (e.g. https://github.com/apache/kafka/pull/11447/files#diff-ad3728ff5f32ed1e88e11e46c0de970bcce01c39767fd3ae666e4a0bfbd97be3L87-L104), except we also don't check for that condition more than necessary.

guozhangwang commented:
> I might be missing the full picture here as I am a bit new to the project, but, as I see it, the algorithm I described above does not suffer from this issue. There, we essentially extract if (record.key() == null || record.value() == null) { code from the operations (e.g. https://github.com/apache/kafka/pull/11447/files#diff-ad3728ff5f32ed1e88e11e46c0de970bcce01c39767fd3ae666e4a0bfbd97be3L87-L104), except we also don't check for that condition more than necessary.

My rationale is that, we do some optimizations at the end of the building phase which may move some repartition up in the topology etc, which would effectively change the topology. And hence, we would not know if all of the downstream operators after the repartition would be filtering null keys until the optimization that changed the topology is done.

Gerrrr commented:
> My rationale is that, we do some optimizations at the end of the building phase which may move some repartition up in the topology etc, which would effectively change the topology. And hence, we would not know if all of the downstream operators after the repartition would be filtering null keys until the optimization that changed the topology is done.

Ah, I missed that! Let me do another iteration on this patch then to make sure that it does not get in the way of suggested optimization.
